doc/man/man1/sign.1
.\" Automatically generated by Pandoc 1.16.0.2
.\"
.TH "SIGN" "1" "February 29, 2016" "Virgil Security CLI (2.0.0)" "Virgil"
.hy
.SH NAME
.PP
\f[B]sign\f[] \[en] sign data
.SH SYNOPSIS
.IP
.nf
\f[C]
\ \ \ \ virgil\ sign\ \ [\-i\ <file>]\ [\-o\ <file>]\ \-k\ <file>\ [\-p\ <arg>]\ [\-V]\ [\-\-]\ [\-\-version]
\ \ \ \ \ \ \ \ \ [\-h]
\f[]
.fi
.SH DESCRIPTION
.PP
Sign data with given user's Private Key.
.SH OPTIONS
.IP
.nf
\f[C]
\ \ \ \ \-i\ <file>,\ \ \-\-in\ <file>
\ \ \ \ \ Data\ to\ be\ signed.\ If\ omitted,\ stdin\ is\ used.

\ \ \ \ \-o\ <file>,\ \ \-\-out\ <file>
\ \ \ \ \ Digest\ sign.\ If\ omitted,\ stdout\ is\ used.

\ \ \ \ \-k\ <file>,\ \ \-\-key\ <file>
\ \ \ \ \ (required)\ \ Signer\[aq]s\ Private\ Key.

\ \ \ \ \-p\ <arg>,\ \ \-\-private\-key\-password\ <arg>
\ \ \ \ \ Password\ to\ be\ used\ for\ Private\ Key\ encryption.

\ \ \ \ \-V,\ \ \-\-VERBOSE
\ \ \ \ \ Show\ detailed\ information

\ \ \ \ \-\-,\ \ \-\-ignore_rest
\ \ \ \ \ Ignores\ the\ rest\ of\ the\ labeled\ arguments\ following\ this\ flag.

\ \ \ \ \-\-version
\ \ \ \ \ Displays\ version\ information\ and\ exits.

\ \ \ \ \-h,\ \ \-\-help
\ \ \ \ \ Displays\ usage\ information\ and\ exits.
\f[]
.fi
.SH EXAMPLES
.IP
.nf
\f[C]
\ \ \ \ virgil\ sign\ \-i\ plain.txt\ \-o\ plain.txt.sign\ \-k\ alice/private.key

\ \ \ \ virgil\ sign\ \-i\ plain.txt\ \-o\ plain.txt.sign\ \-k\ alice/private.key\ \-p\ STRONGPASS
\f[]
.fi
.SH SEE ALSO
.PP
virgil(1)
.PD 0
.P
.PD
verify(1)
.SH AUTHORS
(c) Virgil Security Inc.


ChangeLog
virgil CLI ChangeLog (Sorted per date)

= Version v2.0.0-rc1 released 2016-05-16

Changes:
  * Use virgil-crypto lib 1.6.0
  * Use virgil-sdk-cpp lib 3.2.0
  * Add config file


= Version 0.3.4 released 2015-08-27

Bugfix:
  * Update underlying virgil crypto library to fixed version 1.0.1.

= Version 0.3.3 released 2015-08-18

Bugfix:
  * Embed MSVC redistributable package installation to the CLI package.

Changes:
  * Add executable folder to the PATH environment variable silently on Windows.

= Version 0.3.2 released 2015-08-17

Bugfix:
  * Fix RSA and EC keys generation.

= Version 0.3.1 released 2015-08-17

Bugfix:
  * Add missing win32 runtime libraries.

= Version 0.3.0 released 2015-08-13

Improvements:
  * Add Markdown documentation.

Changes:
  * Replace certificate term with Virgil Public Key term.
  * Remove options to specify sign's output format.
  * Use C++11 standard.

Features:
  * Add command "keyreg" - register user's public key on the Virgil Public Keys service.
  * Add command "keyget" - get user's public key from the Virgil Public Keys service.
  * Add command "confirm" - send confirmation code to the Virgil Public Keys service.
  * Add command "reconfirm" - resend confirmation code to the user for given user's identifier.

= Version 0.2.0 released 2015-05-15

Bugfix:
  * Use standard constructor of class 'ifstream'.
  * Change parameter 'p' long name from 'arg' to 'pwd'.
  * Add required include statement <cstdlib>.

Improvements:
  * Fix misspelling in the help text.
  * Add "happy path" test.
  * Modify 'encrypt' and 'verify' commands to work with Virgil PKI service thru user's identifier.

Changes:
  * Change parameter 'asn1' to 'der'.
  * Change parameters format for 'encrypt' command.

Features:
  * Add 'verify' command.
  * Add 'adduser' command

= Version 0.1.0 released 2015-05-04

Features:
  * Add command "keygen" - generate private key with given parameters.
  * Add command "key2pub" - extract public key from the private key.
  * Add command "pub2cert" - create certificate from the public key and identifiers.
  * Add command "certinfo" - output detailed information about given certificate.
  * Add command "encrypt" - encrypt data for given recipients which can be defined by certificates and by passwords.
  * Add command "decrypt" - decrypt data for given recipient which can be defined by certificate or by password.
  * Add command "sign" - sign data with private key.
  * Add command "verify" - verify data with certificate.


src/commands/base/sign.cxx
/**
 * Copyright (C) 2015 Virgil Security Inc.
 *
 * Lead Maintainer: Virgil Security Inc. <support@virgilsecurity.com>
 *
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 *     (1) Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *
 *     (2) Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in
 *     the documentation and/or other materials provided with the
 *     distribution.
 *
 *     (3) Neither the name of the copyright holder nor the names of its
 *     contributors may be used to endorse or promote products derived from
 *     this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ''AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
 * IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include <fstream>
#include <iostream>
#include <iterator>
#include <stdexcept>
#include <string>

#include <tclap/CmdLine.h>

#include <virgil/crypto/VirgilByteArray.h>
#include <virgil/crypto/VirgilStreamSigner.h>
#include <virgil/crypto/stream/VirgilStreamDataSource.h>

#include <cli/version.h>
#include <cli/util.h>

namespace vcrypto = virgil::crypto;
namespace vcli = virgil::cli;

#ifdef SPLIT_CLI
#define MAIN main
#else
#define MAIN sign_main
#endif

int MAIN(int argc, char** argv) {
    try {
        std::string description = "Sign data with given user's Private Key.\n";

        std::vector<std::string> examples;
        examples.push_back("virgil sign -i plain.txt -o plain.txt.sign -k alice/private.key\n\n");

        examples.push_back("virgil sign -i plain.txt -o plain.txt.sign -k alice/private.key -p STRONGPASS\n\n");

        std::string descriptionMessage = virgil::cli::getDescriptionMessage(description, examples);

        // Parse arguments.
        TCLAP::CmdLine cmd(descriptionMessage, ' ', virgil::cli_version());

        TCLAP::ValueArg<std::string> inArg("i", "in", "Data to be signed. If omitted, stdin is used.", false, "",
                                           "file");

        TCLAP::ValueArg<std::string> outArg("o", "out", "Digest sign. If omitted, stdout is used.", false, "", "file");

        TCLAP::ValueArg<std::string> privateKeyArg("k", "key", "Signer's Private Key.", true, "", "file");

        TCLAP::ValueArg<std::string> privateKeyPasswordArg(
            "p", "private-key-password", "Password to be used for Private Key encryption.", false, "", "arg");

        TCLAP::SwitchArg verboseArg("V", "VERBOSE", "Show detailed information", false);

        cmd.add(verboseArg);
        cmd.add(privateKeyPasswordArg);
        cmd.add(privateKeyArg);
        cmd.add(outArg);
        cmd.add(inArg);
        cmd.parse(argc, argv);

        // Prepare input
        std::istream* inStream;
        std::ifstream inFile;
        if (inArg.getValue().empty() || inArg.getValue() == "-") {
            inStream = &std::cin;
        } else {
            inFile.open(inArg.getValue(), std::ios::in | std::ios::binary);
            if (!inFile) {
                throw std::invalid_argument("cannot read file: " + inArg.getValue());
            }
            inStream = &inFile;
        }

        // Read private key
        vcrypto::VirgilByteArray privateKey = vcli::readPrivateKey(privateKeyArg.getValue());
        vcrypto::VirgilByteArray privateKeyPassword;
        if (privateKeyPasswordArg.isSet()) {
            privateKeyPassword = vcrypto::str2bytes(privateKeyPasswordArg.getValue());
        } else {
            privateKeyPassword = vcli::setPrivateKeyPass(privateKey);
        }

        // Create signer
        vcrypto::VirgilStreamSigner signer;

        // Sign data
        vcrypto::stream::VirgilStreamDataSource dataSource(*inStream);
        vcrypto::VirgilByteArray sign = signer.sign(dataSource, privateKey, privateKeyPassword);

        // Prepare output. Write sign to the output.
        vcli::writeBytes(outArg.getValue(), sign);

        if (verboseArg.isSet()) {
            std::cout << "File signed" << std::endl;
        }
    } catch (TCLAP::ArgException& exception) {
        std::cerr << "sing. Error: " << exception.error() << " for arg " << exception.argId() << std::endl;
        return EXIT_FAILURE;
    } catch (std::exception& exception) {
        std::cerr << "sign. Error: " << exception.what() << std::endl;
        return EXIT_FAILURE;
    }

    return EXIT_SUCCESS;
}
